src/System.Device.Gpio/System/Device/Gpio/Drivers/LibGpiodDriver.Linux.cs

```diff
@@ -0,0 +1,276 @@
+            if (iterator == null)
+            {
+                throw new IOException($"No any chip available, error code: {Marshal.GetLastWin32Error()}");
+            }
+            _chip = Interop.GetNextChipFromChipIterator(iterator);
+            if (_chip == null)
+            {
+                throw new IOException($"No chip found, error code: {Marshal.GetLastWin32Error()}");
+            }
+            // Freeing other chips opened
+            Interop.FreeChipIteratorNoCloseCurrentChip(iterator);
+        }
+
+        private SafeLineHandle RequestEvent(int pinNumber, PinEventTypes eventType)
+        {
+            int eventSuccess = -1;
+            SafeLineHandle pin = Interop.GetChipLineByOffset(_chip, pinNumber);
+            if (pin != null)
+            {
+                if (eventType == PinEventTypes.Rising || eventType == PinEventTypes.Falling)
+                {
+                    eventSuccess = Interop.RequestBothEdgeEventForLine(pin, $"Listen {pinNumber} for falling edge event");
+                }
+                else
+                {
+                    throw new ArgumentException("Invalid or not supported event type requested");
+                }
+            }
+            if (eventSuccess < 0)
+            {
+                throw new IOException($"Error while requesting event listener for pin {pinNumber}, error code: {Marshal.GetLastWin32Error()}");
+            }
+            return pin;
+        }
+
+        protected internal override void AddCallbackForPinValueChangedEvent(int pinNumber, PinEventTypes eventType, PinChangeEventHandler callback)
+        {
+            if (!_pinNumberToEventHandler.TryGetValue(pinNumber, out LibGpiodDriverEventHandler eventHandler))
+            {
+                if (_pinNumberToSafeLineHandle.TryGetValue(pinNumber, out SafeLineHandle pin))
+                    Interop.ReleaseGpiodLine(pin);
+                eventHandler = new LibGpiodDriverEventHandler(pinNumber, new CancellationTokenSource());
+                _pinNumberToEventHandler.TryAdd(pinNumber, eventHandler);
+                eventHandler.PinHandle = RequestEvent(pinNumber, eventType);
+                InitializeEventDetectionTask(eventHandler);
+            }
+            if (eventType == PinEventTypes.Rising)
+            {
+                eventHandler.ValueRising += callback;
+            }
+            else if (eventType == PinEventTypes.Falling)
+            {
+                eventHandler.ValueFalling += callback;
+            }
+        }
+
+        private void InitializeEventDetectionTask(LibGpiodDriverEventHandler eventHandler)
+        {
+            Task pinListeningThread = Task.Run(() =>
+            {
+                int waitResult = -2;
+                while (!eventHandler.CancellationTokenSource.IsCancellationRequested)
+                {
+                    waitResult = Interop.WaitForEventOnLine(eventHandler.PinHandle);
```

Comment:
Will this block forever? If so, shouldn't we use some sort of timeout so that we can check again if cancelation is requested? The way this is now (and assuming this is blocking), then this task will never get canceled even if you send cancellation in the case where an event is not received.

Reply:
no its not blocking forever, its timed out in each 1 millisecond https://github.com/dotnet/iot/pull/133/files#diff-dc75afeeda145b47dfc2bd34d7f04d21R120